src/sentry/snuba/metrics/extraction.py

```diff
@@ -218,21 +221,110 @@ def is_on_demand_metric_query(
+
+    standard_metrics: bool
+    on_demand_metrics: bool
+
+    @classmethod
+    def neither(cls):
+        return cls(standard_metrics=False, on_demand_metrics=False)
+
+    @classmethod
+    def both(cls):
+        return cls(standard_metrics=True, on_demand_metrics=True)
+
+
+def should_use_on_demand_metrics(
+    dataset: Optional[Union[str, Dataset]], aggregate: str, query: Optional[str]
+) -> bool:
+    """On-demand metrics are used if the aggregate and query are supported by on-demand metrics but not standard"""
 
     if not dataset or Dataset(dataset) != Dataset.PerformanceMetrics:
         return False
 
-    if is_standard_metrics_compatible(dataset, aggregate, query):
-        return False
+    supported_by = get_supported_by(aggregate, query)
 
-    for field in _get_aggregate_fields(aggregate):
-        if not _is_on_demand_supported_field(field):
-            return False
+    return not supported_by.standard_metrics and supported_by.on_demand_metrics
+
+
+def get_supported_by(aggregate: str, query: Optional[str] = ""):
+    aggregate_supported_by = _get_aggregate_supported_by(aggregate)
+    query_supported_by = _get_query_supported_by(query)
+
+    standard_metrics = (
+        aggregate_supported_by.standard_metrics and query_supported_by.standard_metrics
+    )
+    on_demand_metrics = (
+        aggregate_supported_by.on_demand_metrics and query_supported_by.on_demand_metrics
+    )
+
+    return SupportedBy(standard_metrics=standard_metrics, on_demand_metrics=on_demand_metrics)
+
+
+def _get_aggregate_supported_by(aggregate: str) -> SupportedBy:
     try:
-        return _is_on_demand_supported_query(event_search.parse_search_query(query))
+        if is_equation(aggregate):
+            # TODO(Ogi): Implement support for equations
+            return SupportedBy.neither()
+
+        function, args, _ = fields.parse_function(aggregate)
+
+        return _get_fn_supported_by(function, args)
+
+    except InvalidSearchQuery:
+        logger.error(f"Failed to parse aggregate: {aggregate}", exc_info=True)
+
+    return SupportedBy.neither()
+
+
+def _get_fn_supported_by(function: str, args: Sequence[str]) -> SupportedBy:
+    fn_support = _get_fn_support(function)
+    args_support = _get_args_support(args)
+
+    standard_metrics = fn_support.standard_metrics and args_support.standard_metrics
+    on_demand_metrics = fn_support.on_demand_metrics and args_support.on_demand_metrics
+
+    return SupportedBy(standard_metrics=standard_metrics, on_demand_metrics=on_demand_metrics)
+
+
+def _get_fn_support(function: str) -> SupportedBy:
+    on_demand_metrics = False
+
+    if function in _SEARCH_TO_METRIC_AGGREGATES and function in _AGGREGATE_TO_METRIC_TYPE:
+        on_demand_metrics = True
+
+    return SupportedBy(standard_metrics=True, on_demand_metrics=on_demand_metrics)
+
+
+def _get_args_support(args: Sequence[str]) -> SupportedBy:
+    if len(args) == 0:
+        return SupportedBy.both()
```

Comment:
There is something wrong, or something that I'm missing.

Are saying here that all functions are supported by standard_metrics but not by on_demand_metrics ?

Reply:
Yes, for example `failure_rate()` is not supported by on-demand. So in case that we have:
aggregate: `failure_rate()` -> supported by standard, but not on-demand
query: `transaction.duration:>0` -> supported by on-demand but not standard
we actually have to fall back to indexed